### PR TITLE
Refactor the expression parser

### PR DESCRIPTION
**Pull request by kyouko-taiga**

### CODE REVIEW

File: Sources/Compiler/Parse/Parser.swift [hunk 1]
```diff
@@ -1538,562 +1531,1066 @@ public enum Parser {
...
-        try state.ast.insert(wellFormed: IntegerLiteralExpr(
-          value: state.lexer.source[token.origin].filter({ $0 != "_" }),
-          origin: token.origin))
-      })
-  )
+    switch head.kind {
+    case .bool:
+      // Boolean literal.
+      _ = state.take()
+      let expr = try state.ast.insert(
```
Comment (dabrahams) [inline Sources/Compiler/Parse/Parser.swift:1757]: Lots of boilerplate repetition in this hunk.  Can we do something generic parameterized on an expr type?

File: Sources/Compiler/Parse/Parser.swift [hunk 2]
```diff
@@ -1538,562 +1531,1066 @@ public enum Parser {
...
-    take(.string)
-      .map({ (state, token) -> NodeID<StringLiteralExpr> in
-        try state.ast.insert(wellFormed: StringLiteralExpr(
-          value: String(state.lexer.source[token.origin].dropFirst().dropLast()),
-          origin: token.origin))
-      })
-  )
+    case .any:
+      // Existential type expression.
+      return try parseExistentialTypeExpr(in: &state).map(AnyExprID.init)
```
Comment (dabrahams) [inline Sources/Compiler/Parse/Parser.swift:1804]: Nice; it's getting smaller, but still looks like this same pattern is repeated over and over so it'd be nice to factor it out.

File: Sources/Compiler/Parse/Parser.swift [hunk 3]
```diff
@@ -1538,562 +1531,1066 @@ public enum Parser {
...
-          origin: state.ast[tree.0.0].origin!.extended(upTo: state.currentIndex)))
-      })
-  )
+  private static func parseNameExprComponent(
+    in state: inout ParserState
+  ) throws -> NameExprComponent? {
+    // Parse the name of the component.
+    guard let name = try parseEntityName(in: &state) else { return nil }
+
+    // If the next token is a left angle bracket, without any leading whitespace, parse a static
```
Comment (dabrahams) [inline Sources/Compiler/Parse/Parser.swift:1904]:
```suggestion
    // If the next token is a left angle bracket without any leading whitespace, parse a static
```

File: Sources/Compiler/Parse/Parser.swift [hunk 4]
```diff
@@ -3033,6 +3319,22 @@ public enum Parser {
...
   }
 
+  private static func expected(
+    _ closerDescription: String,
+    matching opener: Token,
+    in state: ParserState
+  ) -> Diagnostic {
+    expected(
+      "'\(closerDescription)'",
+      at: state.currentLocation,
```
Comment (dabrahams) [inline Sources/Compiler/Parse/Parser.swift:3329]: This is a great example why the "if you split the list everything gets its own line" pattern doesn't work so well..  You end up with these tall columns and things that are closely associated, like `x` and `y` parameters or phrases

```swift
   expected(
      "'\(closerDescription)'", at: state.currentLocation,
```

lose their visible associations.

File: Sources/Compiler/Parse/Parser.swift [hunk 5]
```diff
@@ -3033,6 +3319,22 @@ public enum Parser {
     .error("expected \(subject)", range: location ..< location, children: children)
   }
 
+  private static func expected(
+    _ closerDescription: String,
+    matching opener: Token,
+    in state: ParserState
+  ) -> Diagnostic {
```
Comment (dabrahams) [inline Sources/Compiler/Parse/Parser.swift:3326]:
```suggestion
  private static func expected(_ closerName: String,  matching opener: Token, in state: ParserState) -> Diagnostic {
```

or

```suggestion
  private static func expected(
    _ closerDescription: String,  matching opener: Token, in state: ParserState) -> Diagnostic
  {
```

File: Sources/Compiler/Parse/Parser.swift [hunk 6]
```diff
@@ -3289,6 +3605,25 @@ where Base.Context == ParserState
...
+  init?(_ token: Token) {
+    switch token.kind {
+    case .infix:
+      self = .infix
+    case .prefix:
+      self = .prefix
+    case .postfix:
+      self = .postfix
+    default:
+      return nil
```
Comment (dabrahams) [inline Sources/Compiler/Parse/Parser.swift:3621]:
```suggestion
    case .infix: self = .infix
    case .prefix: self = .prefix
    case .postfix: self = .postfix
    default: return nil
```

File: Sources/Compiler/Parse/ParserState.swift [hunk 7]
```diff
@@ -54,6 +54,9 @@ struct ParserState {
     self.currentIndex = lexer.source.contents.startIndex
   }
 
+  /// The character stream from the current index.
```
Comment (dabrahams) [inline Sources/Compiler/Parse/ParserState.swift:57]: Can you use words other than “character stream” to describe this thing called `characterStream`?  Or better yet, can you find a name for it that doesn't simply describe its notional type?  Is this the `remainingText` or the `unlexedSource` or something?

File: Sources/Compiler/Parse/ParserState.swift [hunk 8]
```diff
@@ -54,6 +54,9 @@ struct ParserState {
...
+  var characterStream: Substring { lexer.source.contents[currentIndex...] }
+
   /// Indicates whether the parser is at global scope.
   var atGlobalScope: Bool { atModuleScope || atNamespaceScope }
```
Comment (dabrahams) [inline Sources/Compiler/Parse/ParserState.swift:61]:
```suggestion
  var isAtGlobalScope: Bool { isAtModuleScope || isAtNamespaceScope }
```

I realize you didn't do anything with this in this PR, but these names don't seem to follow the guidelines.  Also, I think this definition of “global” is a little surprising.  Are static members of types "at global scope?"  if not, why not.

File: Sources/Compiler/Parse/ParserState.swift [hunk 9]
```diff
@@ -135,6 +138,20 @@ struct ParserState {
     return token
   }
 
+  /// Returns whether the next token has the specified kind, or `false` if the stream is empty.
```
Comment (dabrahams) [inline Sources/Compiler/Parse/ParserState.swift:141]:
```suggestion
  /// Returns whether a token of the given `kind` is next in the input.
```

"specified" is just a 3-syllable way to say “given” and even that should usually be avoided.

[And it seems I'll never be comfortable reading “Returns whether” in documentation.  Grumpy old man!]

File: Sources/Compiler/Parse/ParserState.swift [hunk 10]
```diff
@@ -135,6 +138,20 @@ struct ParserState {
...
+  mutating func isNext(_ kind: Token.Kind) -> Bool {
+    peek()?.kind == kind
+  }
+
+  /// Returns whether the next token satisfies `predicate`, or `false` if the stream is empty.
```
Comment (dabrahams) [inline Sources/Compiler/Parse/ParserState.swift:146]:
```suggestion
  /// Returns whether a token satisfying `predicate` is next in the input.
```

File: Sources/Compiler/Parse/ParserState.swift [hunk 11]
```diff
@@ -149,7 +166,7 @@ struct ParserState {
     return token
   }
 
-  /// Consumes and returns the next token, only if it has the specified kind.
+  /// Consumes and returns the next token it has the specified kind.
```
Comment (dabrahams) [inline Sources/Compiler/Parse/ParserState.swift:169]:
```suggestion
  /// Consumes and returns the next token if it has the specified kind.
```

File: Sources/Compiler/Parse/ParserState.swift [hunk 12]
```diff
@@ -167,7 +184,7 @@ struct ParserState {
     return take(kind)
   }
 
-  /// Consumes and returns the next token, if it has the specified predicate.
+  /// Consumes and returns the next token if it satisfies the specified predicate.
```
Comment (dabrahams) [inline Sources/Compiler/Parse/ParserState.swift:187]:
```suggestion
  /// Consumes and returns the next token if it satisfies `predicate`.
```

File: Sources/Compiler/Parse/Token.swift [hunk 13]
```diff
@@ -112,20 +112,24 @@ public struct Token {
 
   /// Indicates whether `self` may be in an operator
   ///
-  /// Use this property rather than testing `self.kind == .oper` as some operators (e.g., `==`) do
-  /// not have the kind `.oper`.
-  public var isOperatorToken: Bool {
+  /// Operators may be represented by one or more tokens of different kinds. For example, `<<` is
```
Comment (dabrahams) [inline Sources/Compiler/Parse/Token.swift:115]:
```suggestion
  /// Operators may be represented by one or more tokens. For example, `<<` is
```

File: Sources/Compiler/TypeChecking/CaptureCollector.swift [hunk 14]
```diff
@@ -559,12 +559,10 @@ struct CaptureCollector {
     into captures: inout FreeSet
   ) {
     switch ast[id].domain {
-    case .none:
+    case .none, .implicit:
```
Comment (dabrahams) [inline Sources/Compiler/TypeChecking/CaptureCollector.swift:562]: Intended functionality change?

**Review by dabrahams — APPROVED** [review]
Looks good overall.  A couple of small things you should probably address.

File: Sources/Compiler/Parse/ParserState.swift (hunk 8 above)
Comment (kyouko-taiga) [reply Sources/Compiler/Parse/ParserState.swift:61]: That property indicates whether the parser is parsing at the global scope, which does not imply that things parsed in type scope can't end up being considered "global" (see `Program.isGlobal(_:)`).

File: Sources/Compiler/Parse/Parser.swift (hunk 5 above)
Comment (kyouko-taiga) [reply Sources/Compiler/Parse/Parser.swift:3326]: Fixed in another PR.

File: Sources/Compiler/TypeChecking/CaptureCollector.swift (hunk 14 above)
Comment (kyouko-taiga) [reply Sources/Compiler/TypeChecking/CaptureCollector.swift:562]: Yep. Previous version was a mistake.

**Pull request merged**